src/Hal/JsonSchema/SchemaFactory.php

```diff
@@ -0,0 +1,132 @@
+            $schemaFactory->addDistinctFormat('jsonhal');
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildSchema(string $className, string $format = 'jsonhal', string $type = Schema::TYPE_OUTPUT, ?string $operationType = null, ?string $operationName = null, ?Schema $schema = null, ?array $serializerContext = null, bool $forceCollection = false): Schema
+    {
+        $schema = $this->schemaFactory->buildSchema($className, $format, $type, $operationType, $operationName, $schema, $serializerContext, $forceCollection);
+        if ('jsonhal' !== $format) {
+            return $schema;
+        }
+
+        $definitions = $schema->getDefinitions();
+        if ($key = $schema->getRootDefinitionKey()) {
+            $definitions[$key]['properties'] = self::BASE_PROPS + ($definitions[$key]['properties'] ?? []);
+
+            return $schema;
+        }
+        if ($key = $schema->getItemsDefinitionKey()) {
+            $definitions[$key]['properties'] = self::BASE_PROPS + ($definitions[$key]['properties'] ?? []);
+        }
+
+        if (($schema['type'] ?? '') === 'array') {
+            $items = $schema['items'];
+            unset($schema['items']);
+
+            $schema['type'] = 'object';
+            $schema['properties'] = [
+                '_embedded' => [
+                    'type' => 'array',
+                    'items' => $items,
+                ],
+                'totalItems' => [
+                    'type' => 'integer',
+                    'minimum' => 0,
+                ],
+                'itemsPerPage' => [
+                    'type' => 'integer',
+                    'minimum' => 0,
+                ],
+                '_links' => [
+                    'type' => 'object',
+                    'properties' => [
+                        'self' => [
+                            'type' => 'object',
+                            'properties' => self::HREF_PROP,
+                        ],
+                        'first' => [
+                            'type' => 'object',
+                            'properties' => self::HREF_PROP,
+                        ],
+                        'last' => [
+                            'type' => 'object',
+                            'properties' => self::HREF_PROP,
+                        ],
+                        'next' => [
+                            'type' => 'object',
+                            'properties' => self::HREF_PROP,
+                        ],
+                        'previous' => [
+                            'type' => 'object',
+                            'properties' => self::HREF_PROP,
+                        ],
+                    ],
+                ],
+            ];
+            $schema['required'] = [
+                '_links',
+                '_embedded',
+            ];
+
+            return $schema;
+        }
+
+        return $schema;
+    }
```

Comment:
Shouldn't you add a public `addDistinctFormat` method here too then?
In case this factory would be decorated too in the future?

Reply:
Good point, will add it.